File: Javascript/16. 3Sum Closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    let n = nums.length
    nums.sort((a,b) => a-b)
    let result = Infinity

    for (let i = 0; i < n-2; i++) {
        if (i > 0 && nums[i] == nums[i-1]) continue;

        let left = i + 1
        let right = n - 1
        while (left < right) {
            let total = nums[i] + nums[left] + nums[right]
            
            if (Math.abs(result-target) > Math.abs(total-target)) {
                result = total
            }
            if (total > target) {
                right--
            } else if (total < target) {
                left++
            } else {
                return target
            }
        }
    }
    return result
};

File: Python/16. 3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        nums.sort()
        result = inf
        
        for i in range(n):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            left = i + 1
            right = n - 1            
            while left < right:
                total = nums[i] + nums[left] + nums[right]
                if abs(result-target) > abs(total-target):
                    result = total
                if total > target:
                    right -= 1
                elif total < target:
                    left += 1
                else:
                    return target
                # print(result)
        return result
        